x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java

```diff
@@ -344,8 +345,30 @@ private void doSaveState(TransformState state, ActionListener<Void> listener) {
                     listener.onResponse(null);
                 }
             }, statsExc -> {
-                logger.error(new ParameterizedMessage("[{}] updating stats of transform failed.", transformConfig.getId()), statsExc);
-                auditor.warning(getJobId(), "Failure updating stats of transform: " + statsExc.getMessage());
+                if (org.elasticsearch.ExceptionsHelper.unwrapCause(statsExc) instanceof VersionConflictEngineException) {
+                    // this should never happen, but indicates a race condition in state persistence:
+                    // - there should be only 1 save persistence at a time
+                    // - this is not a catastrophic failure, resetting the version self-heals on next persistence
+                    // - for tests fail(assert), so we can debug the problem
+                    logger.error(
+                        new ParameterizedMessage(
+                            "[{}] updating stats of transform failed, unexpected version conflict of internal state, resetting to recover.",
+                            transformConfig.getId()
+                        ),
+                        statsExc
+                    );
+                    auditor.warning(
+                        getJobId(),
+                        "Failure updating stats of transform, unexpected version conflict of internal state, resetting to recover: "
+                            + statsExc.getMessage()
+                    );
+                    assert false : "[" + getJobId() + "] updating stats of transform failed, unexpected version conflict of internal state";
+                    // the version must be reset, otherwise it will never succeed for this indexer
+                    this.seqNoPrimaryTermAndIndex.set(null);
```

Comment:
Should we update `VersionConflictEngineException` so that it can provide the current term and seqno? It seems preferable to set this to what the current write values are instead of `null` to me.

Reply:
I had a similar thought, but I think its intentionally not exposed.
The use case are concurrent document updates, the app retrieves the current document, alters a field and pushes the document back. If a concurrent update happened meanwhile, the version conflict ensures consistency, so you can't undo/overwrite the update that happened in another thread/process/whatever. The correct handling in case the version conflict fires: the app should re-read the doc (including the current `seq_no`), re-do the change and push it again with the current `seq_no`.

That means, setting the internal version to `null` is actually incorrect. My problem: If I re-read the state document: how should I merge the current state with the state I think is correct? I do not have a diff.

I will think about it a bit more. The situation should not happen in the 1st place, still I think we should handle the problem and not fail endlessly. But setting the transform to `FAILED` seems wrong to me, too.